distributed/scheduler.py

```diff
@@ -4893,49 +4893,39 @@ async def remove_worker_from_events():
 
         return "OK"
 
-    async def stimulus_cancel(self, keys, client, force=False):
+    def stimulus_cancel(
+        self, keys: Sequence[str], client: str, force: bool = False
+    ) -> None:
         """Stop execution on a list of keys"""
         logger.info("Client %s requests to cancel %d keys", client, len(keys))
         if client:
             self.log_event(
                 client, {"action": "cancel", "count": len(keys), "force": force}
             )
-
-        await asyncio.gather(
-            *[self._cancel_key(key, client, force=force) for key in keys]
-        )
-
-    async def _cancel_key(self, key, client, force=False):
-        """Cancel a particular key and all dependents"""
-        # TODO: this should be converted to use the transition mechanism
-        ts = self.tasks.get(key)
-        try:
-            cs = self.clients[client]
-        except KeyError:
-            return
-
-        # no key yet, lets try again in a moment
-        start = time()
-        while ts is None or not ts.who_wants:
-            await asyncio.sleep(0.1)
-            ts = self.tasks.get(key)
-            if time() - start >= 1:
+        report = list()
+        clients = list()
+        for key in keys:
+            ts: TaskState | None = self.tasks.get(key)
+            if not ts:
+                continue
+            try:
+                cs: ClientState = self.clients[client]
+            except KeyError:
                 return
 
-        if force or ts.who_wants == {cs}:  # no one else wants this key
-            await asyncio.gather(
-                *[
-                    self._cancel_key(dts.key, client, force=force)
-                    for dts in ts.dependents
-                ]
-            )
-            logger.info("Scheduler cancels key %s.  Force=%s", key, force)
-            self.report({"op": "cancelled-key", "key": key})
-        clients = list(ts.who_wants) if force else [cs]
+            if force or ts.who_wants == {cs}:  # no one else wants this key
+                if ts.dependents:
+                    self.stimulus_cancel(
+                        [dts.key for dts in ts.dependents], client, force=force
+                    )
+                logger.info("Scheduler cancels key %s.  Force=%s", key, force)
+                report.append(key)
+            clients.extend(list(ts.who_wants) if force else [cs])
         for cs in clients:
             self.client_releases_keys(
-                keys=[key], client=cs.client_key, stimulus_id=f"cancel-key-{time()}"
+                keys=report, client=cs.client_key, stimulus_id=f"cancel-key-{time()}"
```

Comment:
We now ask all clients waiting for at least one key to release all of the keys. Should we make this smarter to reduce the network traffic?

Reply:
Ignore that, we filter the keys in `_client_releases_keys` before sending anything.